lib/writers/git.go

```diff
@@ -406,3 +443,129 @@ func sshUsernameFromURL(url string) (string, error) {
 	}
 	return ep.User, nil
 }
+
+// for unit tests
+
+var GenerateGitHubAppJWT = generateGitHubAppJWT
+var GetGitHubInstallationTokenWithExpiry = getGitHubInstallationTokenWithExpiry
+var StartGitHubTokenAutoRefresh = startGitHubTokenAutoRefresh
+
+// startGitHubTokenAutoRefresh starts a goroutine that refreshes the GitHub installation token
+func startGitHubTokenAutoRefresh(
+	logger logr.Logger,
+	ba *githttp.BasicAuth,
+	appID, installationID, privateKey string,
+	initialExpiry time.Time,
+) {
+	const skew = 5 * time.Minute
+	const minSleep = 30 * time.Second
+	const retryBackoff = 30 * time.Second
+
+	go func() {
+		expiry := initialExpiry.UTC()
+
+		for {
+			wait := time.Until(expiry.Add(-skew))
+			if wait < minSleep {
+				wait = minSleep
+			}
+			timer := time.NewTimer(wait)
+			<-timer.C
+
+			j, err := GenerateGitHubAppJWT(appID, privateKey)
+			if err != nil {
+				logger.Error(err, "github app auth: failed to generate JWT, retrying")
+				time.Sleep(retryBackoff)
+				continue
+			}
+
+			tok, exp, err := GetGitHubInstallationTokenWithExpiry(installationID, j)
+			if err != nil {
+				logger.Error(err, "github app auth: failed to fetch installation token, retrying")
+				time.Sleep(retryBackoff)
+				continue
+			}
+
+			ba.Password = tok
+			expiry = exp.UTC()
+
+			logger.Info("github app auth: refreshed installation token", "expiresAt", expiry.Format(time.RFC3339))
+		}
+	}()
+}
+
+// generateGitHubAppJWT creates a signed JWT for GitHub App authentication
+func generateGitHubAppJWT(appID string, privateKey string) (string, error) {
```

Comment:
nit: any reason to handcraft both this and the next methods vs using a github sdk, like https://github.com/octokit/go-sdk?

Reply:
There's a lot of complexity in these two functions and they are not tested. Could we either: use the sdk or write some tests?

Reply:
octokit/go-sdk only provides end-to-end clients, doesn't provide token generation functionality. So, I kept the implementation, but added new tests. 👍